types/node/v6/index.d.ts

```diff
@@ -2596,7 +2597,8 @@ declare module "fs" {
      */
     export function mkdtempSync(prefix: string): string;
     export function readdir(path: string | Buffer, callback?: (err: NodeJS.ErrnoException, files: string[]) => void): void;
-    export function readdirSync(path: string | Buffer): string[];
+    export function readdir(path: string | Buffer, opts: string | {encoding?: string}, callback?: (err: NodeJS.ErrnoException, files: string[]) => void): void;
+    export function readdirSync(path: string | Buffer, opts?: string | {encoding?: string}): string[];
```

Comment:
Depending on `encoding` the result may be also a `Buffer[]`.

Reply:
@Flarna c15ee71643761c9f6e63cc171f0cbb6d4a3296bd should do it

Reply:
Unfortunately not. Typescript chooses the first matching overload - in your case that one with `string` even `"buffer"` is given therefore `fs.readdirSync("path", { encoding: "buffer" })` will return `string[]`.

If the encoding passed is just a string with unknown content for typescript (e.a. passed in from somewhere) the result should be `string[] | Buffer[]`. Only if the string is known to typescript it can nail it down to either `string[]` or `Buffer[]`. See nodejs v9 typings for details.

Reply:
Ugh, good spot. Copied over the types from v9. Just need to wait until the `esModuleInterop` issue is fixed on master so CI can pass.